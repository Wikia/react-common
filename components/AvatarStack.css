/**
* Implemenation taken from http://www.antimath.info/css/sass-sqrt-function/
*/
/**
* Implementation taken from http://alienryderflex.com/hsp.html
*/
/**
* Implemenation taken from http://www.antimath.info/css/sass-sqrt-function/
*/
/**
* Implementation taken from http://alienryderflex.com/hsp.html
*/
/**
* Implemenation taken from http://www.antimath.info/css/sass-sqrt-function/
*/
/**
* Implementation taken from http://alienryderflex.com/hsp.html
*/
/**
* Implemenation taken from http://www.antimath.info/css/sass-sqrt-function/
*/
/**
* Implementation taken from http://alienryderflex.com/hsp.html
*/
/**
* Implemenation taken from http://www.antimath.info/css/sass-sqrt-function/
*/
/**
* Implementation taken from http://alienryderflex.com/hsp.html
*/
.wds-avatar-stack {
  align-items: center;
  display: flex;
  margin: 8px 18px 6px 40px;
}
.wds-avatar-stack__overflow {
  border: 0;
  font-size: 12px;
  padding-left: 6px;
  width: auto;
}
.wds-avatar-stack .wds-avatar:nth-child(1) {
  position: relative;
  z-index: 6;
}
.wds-avatar-stack .wds-avatar:nth-child(2) {
  position: relative;
  z-index: 5;
  margin-left: -10px;
}
.wds-avatar-stack .wds-avatar:nth-child(3) {
  position: relative;
  z-index: 4;
  margin-left: -10px;
}
.wds-avatar-stack .wds-avatar:nth-child(4) {
  position: relative;
  z-index: 3;
  margin-left: -10px;
}
.wds-avatar-stack .wds-avatar:nth-child(5) {
  position: relative;
  z-index: 2;
  margin-left: -10px;
}
.wds-avatar-stack .wds-avatar:nth-child(6) {
  position: relative;
  z-index: 1;
  margin-left: -10px;
}