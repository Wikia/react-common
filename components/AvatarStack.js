'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = _interopDefault(require('react'));
var PropTypes = _interopDefault(require('prop-types'));
var classNames = _interopDefault(require('classnames'));
var AvatarBadgeAdmin = _interopDefault(require('@wikia/react-common/assets/AvatarBadgeAdmin'));
var AvatarBadgeContentModerator = _interopDefault(require('@wikia/react-common/assets/AvatarBadgeContentModerator'));
var AvatarBadgeDiscussionModerator = _interopDefault(require('@wikia/react-common/assets/AvatarBadgeDiscussionModerator'));
var AvatarBadgeGlobalDiscussionsModerator = _interopDefault(require('@wikia/react-common/assets/AvatarBadgeGlobalDiscussionsModerator'));
var AvatarBadgeHelper = _interopDefault(require('@wikia/react-common/assets/AvatarBadgeHelper'));
var AvatarBadgeStaff = _interopDefault(require('@wikia/react-common/assets/AvatarBadgeStaff'));
var AvatarBadgeVstf = _interopDefault(require('@wikia/react-common/assets/AvatarBadgeVstf'));
var IconAvatar = _interopDefault(require('@wikia/react-common/icons/IconAvatar'));

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function");
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      writable: true,
      configurable: true
    }
  });
  if (superClass) _setPrototypeOf(subClass, superClass);
}

function _getPrototypeOf(o) {
  _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
    return o.__proto__ || Object.getPrototypeOf(o);
  };
  return _getPrototypeOf(o);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf(o, p);
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

function _possibleConstructorReturn(self, call) {
  if (call && (typeof call === "object" || typeof call === "function")) {
    return call;
  }

  return _assertThisInitialized(self);
}

var Badge = function Badge(_ref) {
  var badge = _ref.badge,
      className = _ref.className;
  var badgeIcons = {
    sysop: {
      icon: React.createElement(AvatarBadgeAdmin, null),
      title: 'Community Admin'
    },
    'content-moderator': {
      icon: React.createElement(AvatarBadgeContentModerator, null),
      title: 'Content Moderator'
    },
    threadmoderator: {
      icon: React.createElement(AvatarBadgeDiscussionModerator, null),
      title: 'Discussions Moderator'
    },
    'global-discussions-moderator': {
      icon: React.createElement(AvatarBadgeGlobalDiscussionsModerator, null),
      title: 'Global Discussions Moderator'
    },
    helper: {
      icon: React.createElement(AvatarBadgeHelper, null),
      title: 'FANDOM Helper'
    },
    staff: {
      icon: React.createElement(AvatarBadgeStaff, null),
      title: 'FANDOM Staff'
    },
    vstf: {
      icon: React.createElement(AvatarBadgeVstf, null),
      title: 'VSTF'
    }
  };
  var badgeIcon = badgeIcons[badge];
  return badgeIcon ? React.createElement("span", {
    title: badgeIcon.title,
    className: "wds-avatar__badge ".concat(className)
  }, badgeIcon.icon) : null;
};

Badge.propTypes = {
  badge: PropTypes.string,
  className: PropTypes.string
};
Badge.defaultProps = {
  badge: undefined,
  className: ''
};

var AvatarImage = function AvatarImage(_ref) {
  var src = _ref.src,
      alt = _ref.alt;
  var className = 'wds-avatar__image';
  return src ? React.createElement("img", {
    src: src,
    alt: alt,
    title: alt,
    className: className
  }) : React.createElement(IconAvatar, {
    className: "wds-avatar__image",
    title: className
  });
};

AvatarImage.propTypes = {
  alt: PropTypes.string,
  src: PropTypes.string
};
AvatarImage.defaultProps = {
  alt: undefined,
  src: undefined
};

function getAvatarImage(href, alt, src, linkBuilder) {
  var avatarImage = React.createElement(AvatarImage, {
    alt: alt,
    src: src
  });

  if (linkBuilder) {
    return linkBuilder(avatarImage);
  }

  return href ? React.createElement("a", {
    href: href
  }, avatarImage) : avatarImage;
}

function _createSuper(Derived) {
  function isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;

    try {
      Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
      return true;
    } catch (e) {
      return false;
    }
  }

  return function () {
    var Super = _getPrototypeOf(Derived),
        result;

    if (isNativeReflectConstruct()) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}

var Avatar = /*#__PURE__*/function (_React$Component) {
  _inherits(Avatar, _React$Component);

  var _super = _createSuper(Avatar);

  function Avatar() {
    var _this;

    _classCallCheck(this, Avatar);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      imageSrc: _this.props.src
    });

    return _this;
  }

  _createClass(Avatar, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      if (!this.state.imageSrc && this.props.userId) {
        // eslint-disable-next-line
        fetch("https://services.wikia.com/user-attribute/user/".concat(this.props.userId, "/attr/avatar")).then(function (response) {
          return response.json();
        }).then(function (data) {
          return _this2.setState({
            imageSrc: data.value
          });
        });
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          alt = _this$props.alt,
          badge = _this$props.badge,
          className = _this$props.className,
          href = _this$props.href,
          title = _this$props.title,
          linkBuilder = _this$props.linkBuilder;
      return React.createElement("div", {
        className: classNames('wds-avatar', className),
        title: title
      }, getAvatarImage(href, alt, this.state.imageSrc, linkBuilder), badge && React.createElement(Badge, {
        badge: badge
      }));
    }
  }]);

  return Avatar;
}(React.Component);

Avatar.propTypes = {
  /** Alt text for avatar */
  alt: PropTypes.string,

  /** Badge to display for avatar. */
  badge: PropTypes.oneOf(['content-moderator', 'discussion-moderator', 'sysop', 'global-discussions-moderator', 'helper', 'staff', 'vstf', '']),

  /** Additional class name */
  className: PropTypes.string,

  /** Link to user's profile */
  href: PropTypes.string,

  /** Function which returns wrapped AvatarImage, accepts one argument avatarImage which provides the actual avatar image */
  linkBuilder: PropTypes.func,

  /** Image src for avatar */
  src: PropTypes.string,

  /** Title attribute for avatar */
  title: PropTypes.string,

  /** Optional user ID to fetch avatar for, if `src` prop is not provided */
  userId: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};
Avatar.defaultProps = {
  alt: 'User avatar',
  badge: undefined,
  className: undefined,
  href: undefined,
  linkBuilder: undefined,
  src: undefined,
  title: undefined,
  userId: undefined
};

/* eslint-disable react/no-array-index-key */

var AvatarStack = function AvatarStack(_ref) {
  var avatars = _ref.avatars,
      overrideCount = _ref.overrideCount,
      maxStackSize = _ref.maxStackSize,
      hideOverflow = _ref.hideOverflow,
      className = _ref.className;
  var count = overrideCount || avatars.length;
  var overflow = !hideOverflow && (count > maxStackSize ? count - maxStackSize : 0);
  return React.createElement("div", {
    className: classNames('wds-avatar-stack', className)
  }, avatars.slice(0, maxStackSize).map(function (avatarProps, index) {
    return React.createElement(Avatar, _extends({
      key: index
    }, avatarProps));
  }), overflow && React.createElement("div", {
    className: "wds-avatar-stack__overflow"
  }, "+".concat(overflow)));
};

AvatarStack.propTypes = {
  /* An array of `Avatar` props */
  avatars: PropTypes.arrayOf(PropTypes.shape()).isRequired,

  /* Additional class name */
  className: PropTypes.string,

  /* Flag to hide overflow */
  hideOverflow: PropTypes.bool,

  /* Max stack size */
  maxStackSize: PropTypes.number,

  /* If specified, it will be used instead of `avatars.length` */
  overrideCount: PropTypes.number
};
AvatarStack.defaultProps = {
  className: '',
  hideOverflow: false,
  maxStackSize: 5,
  overrideCount: 0
};

module.exports = AvatarStack;
