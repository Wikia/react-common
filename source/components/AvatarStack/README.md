Defaults:
```js
<AvatarStack
    avatars={[{ src: 'https://static.wikia.nocookie.net/438f6e23-1a0f-4351-a34d-6f8fa734a246', alt: 'Image', link: 'some://valid.url', badge: 'admin' }, { src: 'https://static.wikia.nocookie.net/438f6e23-1a0f-4351-a34d-6f8fa734a246', alt: 'Image', link: 'some://valid.url', badge: 'staff' }, { src: 'https://static.wikia.nocookie.net/fcdbab4c-8838-487e-89fd-5f0782974edf', alt: 'Image', link: 'some://valid.url', badge: 'vstf' }]}
/>
```

With maxStackSize

```js
<AvatarStack
    avatars={[{ src: 'https://static.wikia.nocookie.net/438f6e23-1a0f-4351-a34d-6f8fa734a246', alt: 'Image', link: 'some://valid.url', badge: 'admin' }, { src: 'https://static.wikia.nocookie.net/438f6e23-1a0f-4351-a34d-6f8fa734a246', alt: 'Image', link: 'some://valid.url', badge: 'staff' }, { src: 'https://static.wikia.nocookie.net/fcdbab4c-8838-487e-89fd-5f0782974edf', alt: 'Image', link: 'some://valid.url', badge: 'vstf' }]}
    maxStackSize={2}
/>
```

With hidden overflow

```js
<AvatarStack
    avatars={[{ src: 'https://static.wikia.nocookie.net/438f6e23-1a0f-4351-a34d-6f8fa734a246', alt: 'Image', link: 'some://valid.url', badge: 'admin' }, { src: 'https://static.wikia.nocookie.net/438f6e23-1a0f-4351-a34d-6f8fa734a246', alt: 'Image', link: 'some://valid.url', badge: 'staff' }, { src: 'https://static.wikia.nocookie.net/fcdbab4c-8838-487e-89fd-5f0782974edf', alt: 'Image', link: 'some://valid.url', badge: 'vstf' }]}
    maxStackSize={2}
    hideOverflow
/>
```

With overridden counter

```js
<AvatarStack
    avatars={[{ src: 'https://static.wikia.nocookie.net/438f6e23-1a0f-4351-a34d-6f8fa734a246', alt: 'Image', link: 'some://valid.url', badge: 'admin' }, { src: 'https://static.wikia.nocookie.net/438f6e23-1a0f-4351-a34d-6f8fa734a246', alt: 'Image', link: 'some://valid.url', badge: 'staff' }, { src: 'https://static.wikia.nocookie.net/fcdbab4c-8838-487e-89fd-5f0782974edf', alt: 'Image', link: 'some://valid.url', badge: 'vstf' }]}
    maxStackSize={2}
    overrideCount={6}
/>
```
