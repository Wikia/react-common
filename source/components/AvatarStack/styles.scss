@import "~@fandom/design-system/dist/scss/wds-mixins/index.scss";
@import "~@fandom/design-system/dist/scss/wds-variables/index.scss";
@import "~@fandom/design-system/dist/scss/wds-components/_avatar-stack.scss";
