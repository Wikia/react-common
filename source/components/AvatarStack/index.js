import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import Avatar from '../Avatar';

import './styles.scss';

/* eslint-disable react/no-array-index-key */
const AvatarStack = ({ avatars, overrideCount, maxStackSize, hideOverflow, className }) => {
    const count = overrideCount || avatars.length;
    const overflow = !hideOverflow && (count > maxStackSize ? count - maxStackSize : 0);

    return (
        <div className={classNames('wds-avatar-stack', className)}>
            {
                avatars
                    .slice(0, maxStackSize)
                    .map(({ src, alt, link, badge = '' }, index) => (
                        <Avatar key={index} src={src} alt={alt} badge={badge} href={link} />
                    ))
            }
            {
                overflow && (
                    <div className="wds-avatar-stack__overflow">
                        {`+${overflow}`}
                    </div>
                )
            }
        </div>
    );
};

AvatarStack.propTypes = {
    avatars: PropTypes.array.isRequired,
    className: PropTypes.string,
    hideOverflow: PropTypes.bool,
    maxStackSize: PropTypes.number,
    overrideCount: PropTypes.number,
};

AvatarStack.defaultProps = {
    className: '',
    hideOverflow: false,
    maxStackSize: 5,
    overrideCount: 0,
};

export default AvatarStack;
