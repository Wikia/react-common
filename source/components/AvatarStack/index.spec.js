import { shallow } from 'enzyme';
import React from 'react';
import merge from 'lodash/merge';

import AvatarStack from './index';

const defaultProps = {
    avatars: [
        { src: 'some.url', alt: 'User', link: 'some.link', badge: 'sysop' },
        { src: 'some.other.url', alt: 'Other user', link: 'some.other.link' },
    ],
};

function renderComponent(props) {
    const computedProps = merge({}, defaultProps, props);

    return shallow(<AvatarStack {...computedProps} />);
}

test('AvatarStack renders with default props', () => {
    const wrapper = renderComponent();

    expect(wrapper).toMatchSnapshot();
});

test('AvatarStack renders with overrideCount', () => {
    const wrapper = renderComponent({ overrideCount: 1 });

    expect(wrapper).toMatchSnapshot();
});

test('AvatarStack renders when there are less avatars than specified by maxStackSize', () => {
    const wrapper = renderComponent({ maxStackSize: 3 });

    expect(wrapper).toMatchSnapshot();
});

test('AvatarStack renders when there are more avatars than specified by maxStackSize', () => {
    const wrapper = renderComponent({ maxStackSize: 1 });

    expect(wrapper).toMatchSnapshot();
});

test('AvatarStack renders with hideOverflow set to true', () => {
    const wrapper = renderComponent({ hideOverflow: true, maxStackSize: 1 });

    expect(wrapper).toMatchSnapshot();
});
