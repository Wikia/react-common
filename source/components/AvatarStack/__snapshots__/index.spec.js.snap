// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarStack renders when there are less avatars than specified by maxStackSize 1`] = `
<div
  className="wds-avatar-stack"
>
  <Avatar
    alt="User"
    badge="sysop"
    href="some.link"
    key="0"
    src="some.url"
  />
  <Avatar
    alt="Other user"
    badge=""
    href="some.other.link"
    key="1"
    src="some.other.url"
  />
  0
</div>
`;

exports[`AvatarStack renders when there are more avatars than specified by maxStackSize 1`] = `
<div
  className="wds-avatar-stack"
>
  <Avatar
    alt="User"
    badge="sysop"
    href="some.link"
    key="0"
    src="some.url"
  />
  <div
    className="wds-avatar-stack__overflow"
  >
    +1
  </div>
</div>
`;

exports[`AvatarStack renders with default props 1`] = `
<div
  className="wds-avatar-stack"
>
  <Avatar
    alt="User"
    badge="sysop"
    href="some.link"
    key="0"
    src="some.url"
  />
  <Avatar
    alt="Other user"
    badge=""
    href="some.other.link"
    key="1"
    src="some.other.url"
  />
  0
</div>
`;

exports[`AvatarStack renders with hideOverflow set to true 1`] = `
<div
  className="wds-avatar-stack"
>
  <Avatar
    alt="User"
    badge="sysop"
    href="some.link"
    key="0"
    src="some.url"
  />
</div>
`;

exports[`AvatarStack renders with overrideCount 1`] = `
<div
  className="wds-avatar-stack"
>
  <Avatar
    alt="User"
    badge="sysop"
    href="some.link"
    key="0"
    src="some.url"
  />
  <Avatar
    alt="Other user"
    badge=""
    href="some.other.link"
    key="1"
    src="some.other.url"
  />
  0
</div>
`;
